Allow WebFlux ApiVersionResolver to return a Mono

JKaplanEmpty-Nes opened the pull request:
The current WebFlux API resolution returned a nullable String, this prevented custom API resolvers to be based on values which would need to be blocked. 
For example, creating an API version resolver based on the Authorized Principal. Since this is returned as a Mono in a server exchange.

JKaplanEmpty-Nes committed:
Change API version resolution from nullable String to Mono String

Signed-off-by: Jonathan Kaplan <jkaplan@empty-nes.com>

JKaplanEmpty-Nes commented:
> It is unfortunately too late for this type of change. This would break existing public contracts in a spectacular fashion. I will leave this open for now so we can discuss the use case.

What if I changed the public contracts so that the old return was there, but added a new reactive one and by default it would enclose the old result in a mono?

bclozel commented:
We can always consider solutions like that. Let us discuss this matter first.

rstoyanchev reviewed (changes requested):
This was unfortunately an oversight, but we need to consider the best way to get there with the least disruption.

I do like the idea of allowing an `ApiVersionResolver` to be sync or async, so the `resolveVersionReactively` method with a default implementation is good, but I would call it `resolveVersionAsync`.

However, it doesn't make sense to keep both sync and async variants at the level of `ApiVersionStrategy`, so we should deprecate the sync version for removal, and again call the new method `resolveVersionAsync`. A similar pattern for `resolveParseAndValidate`.

During the deprecation phase, we do need to continue to call the sync `resolveParseAndValidate` first, suppressing any `MissingApiVersionException`, and then conditionally call the new async variant if the sync version did not produce a version.

Let me know if you plan to make those updates.

rstoyanchev commented on spring-webflux/src/main/java/org/springframework/web/reactive/accept/ApiVersionStrategy.java:
```diff
@@ -65,27 +81,44 @@ void validateVersion(@Nullable Comparable<?> requestVersion, ServerWebExchange e
+	default Mono<Comparable<?>> resolveParseAndValidateVersion(ServerWebExchange exchange) {
+		return Mono.defer(()-> {
+					try {
+						return Mono.justOrEmpty(this.resolveVersion(exchange));
```
This is a blocking call within a reactive chain with non-blocking assumptions. `defer` doesn't use a different thread, it is just a way to make the code "cold", not triggered until subscription. You need `.subscribeOn` for this.

That said, related to the previous comment, I don't think the blocking version should be called here. Calling both sync and async should be done in AbstractHandlerMapping.

JKaplanEmpty-Nes replied:
Just have the async one called in ApiVersionStrategy?

rstoyanchev replied:
ApiVersionStrategy should have both sync and async versions (same patters as in ApiVersionResolver). The async version can call only the async resolve methods, while the sync calls the sync resolve methods.

rstoyanchev commented on spring-webflux/src/main/java/org/springframework/web/reactive/accept/ApiVersionStrategy.java:
```diff
@@ -59,6 +75,25 @@ public interface ApiVersionStrategy {
+			return Mono.error(ex);
+		}
+	}
+
```
I'm processing this locally now, but wondering is there a a specific reason to expose this an additional, separate, new method?

rstoyanchev commented:
@JKaplanEmpty-Nes as I mentioned before, I'm already making changes locally. Please, don't push any more changes. All of those will not be included.

JKaplanEmpty-Nes commented:
> Yeah, processing as in processing the PR, not digesting it.
> 
> I've merged the changes and also evolved the solution a bit further. The main thing to point out are the changes in `ApiVersionResolver`, deprecating the sync method and making the async method the main one instead (for `@FunctionalInterface`), and creating an `SyncApiVersionResolver` extension. This is a better setup since otherwise if you implement the async method, the sync one remains without an obvious implementation. It also matches a pattern we've used elsewhere in WebFlux. This does break lambda implementations, but that's inevitable with the new setup after the deprecation phase, so it's better to absorb that now. In any case the expectation is for concrete implementations so I think this should have limited impact at this point in time early in the 7.0.x phase.
> 
> Hopefully the rest is self explanatory and makes sense, but feel free to comment.

As long as we have an async resolver then I am absolutely delighted. My first commit also changed the defaults to the asynced version but I changed it to the synced version based on the first suggestion.
